File: 프로그래머스/2/49994. 방문 길이/방문 길이.cpp
#include <string>
#include <vector>
using namespace std;

vector<vector<vector<bool>>> visited;

struct Pos
{
    int x;
    int y;
};

bool CanGo(Pos nPos)
{
    int x = nPos.x;
    int y = nPos.y;
    if (x < -5 || x > 5) return false;
    if (y < -5 || y > 5) return false;
    return true;
}

bool isVisitedBySamePath(Pos bPos, Pos nPos, char dirC)
{
    // 벡터 탐색을 위한 인덱스
    int bx, by;
    int nx, ny;
    int dir = 0;
    if (dirC == 'U')
    {
        dir = 0;
    }
    else if (dirC == 'D')
    {
        dir = 2;
    }
    else if (dirC == 'R')
    {
         dir = 1;
    }
    else if (dirC == 'L')
    {
        dir = 3;
    }
    
    bx = (bPos.x + 5);
    by = (bPos.y + 5);
    nx = (nPos.x + 5);
    ny = (nPos.y + 5);
    
    
    if (visited[by][bx][dir] || visited[ny][nx][(dir + 2)%4]) return true;
    else 
    {
        visited[by][bx][dir] = true;
        return false;
    }
}

void move(char c, Pos& currentpos, int& count)
{
    Pos nPos = currentpos;
    if (c == 'U')
    {
        nPos.y += 1;
    }
    else if (c == 'D')
    {
        nPos.y -= 1;
    }
    else if (c == 'R')
    {
        nPos.x += 1;
    }
    else if (c == 'L')
    {
        nPos.x -= 1;
    }
    
    if (CanGo(nPos))
    {
        if (!isVisitedBySamePath(currentpos, nPos, c)) count++;
        currentpos.x = nPos.x;
        currentpos.y = nPos.y;
    }
}

int solution(string dirs) 
{
    int answer = 0;
    visited.resize(11, vector<vector<bool>>(11, vector<bool>(4, false)));
    
    Pos currentPos = Pos{0,0};
    
    for (auto c : dirs)
    {
        move(c, currentPos ,answer);
    }
    
    return answer;
}

File: 프로그래머스/2/49994. 방문 길이/방문 길이.cs
using System;

public class Solution {

    private bool[,,] visited;
    
    public int solution(string dirs) {
        int answer = 0;
        int cx = 0;
        int cy = 0;
        // x y 방문한 방향
        visited = new bool[12,12,4];
        
        foreach (var dir in dirs){
            int dx = 0;
            int dy = 0;
            int dirVec = 0;
            
            if (dir == 'R') { dx = 1; dy = 0; dirVec = 0;}
            else if (dir == 'D') { dx = 0; dy = -1; dirVec = 1;}
            else if (dir == 'L') { dx = -1; dy = 0; dirVec = 2;}
            else if (dir == 'U') { dx = 0; dy = 1; dirVec = 3;}
            
            int nx = cx + dx;
            int ny = cy + dy;

            // condition check
            if (-5 > nx || nx > 5) continue;
            if (-5 > ny || ny > 5) continue;
            if (visited[nx+5,ny+5,(dirVec + 2) % 4] != true && visited[cx+5,cy+5,(dirVec) % 4] != true)
            {
                visited[nx+5,ny+5,(dirVec + 2) % 4] = true;
                answer++;
            }

            cx = nx;
            cy = ny;
        }
        
        return answer;
    }
}